File: plugin-core/src/main/java/com/github/catatafishen/ideagentforcopilot/settings/CodexClientConfigurable.java
package com.github.catatafishen.ideagentforcopilot.settings;

import com.github.catatafishen.ideagentforcopilot.agent.codex.CodexAppServerClient;
import com.github.catatafishen.ideagentforcopilot.agent.codex.CodexCredentials;
import com.github.catatafishen.ideagentforcopilot.ui.AuthTerminalHelperKt;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.options.Configurable;
import com.intellij.openapi.project.Project;
import com.intellij.ui.HyperlinkLabel;
import com.intellij.ui.components.JBLabel;
import com.intellij.ui.components.JBTextField;
import com.intellij.util.ui.FormBuilder;
import com.intellij.util.ui.JBUI;
import com.intellij.util.ui.UIUtil;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import java.awt.*;

public final class CodexClientConfigurable implements Configurable {

    private static final String CHECKING = "Checking…";

    private final Project project;

    private JBLabel binaryStatusLabel;
    private JBLabel authStatusLabel;
    private JBTextField binaryPathField;
    private JPanel mainPanel;

    public CodexClientConfigurable(@NotNull Project project) {
        this.project = project;
    }

    @Override
    public @Nls(capitalization = Nls.Capitalization.Title) String getDisplayName() {
        return "Codex (Experimental)";
    }

    @Override
    public @NotNull JComponent createComponent() {
        binaryStatusLabel = new JBLabel(CHECKING);
        authStatusLabel = new JBLabel(CHECKING);

        binaryPathField = new JBTextField();
        binaryPathField.getEmptyText().setText("Auto-detect (leave empty)");
        binaryPathField.setToolTipText("Absolute path to the codex binary. Leave empty to find it on PATH.");

        HyperlinkLabel installLink = new HyperlinkLabel("Install Codex CLI — npmjs.com/@openai/codex");
        installLink.setHyperlinkTarget("https://www.npmjs.com/package/@openai/codex");

        JBLabel installNote = new JBLabel(
            "<html>Install with <code>npm install -g @openai/codex</code>, then run <code>codex login</code>.</html>");
        installNote.setForeground(UIUtil.getContextHelpForeground());
        installNote.setFont(JBUI.Fonts.smallFont());

        JButton signInButton = new JButton("Sign In (codex login)");
        signInButton.addActionListener(e -> openSignInTerminal(false));

        JButton signInDeviceButton = new JButton("Sign In — headless (codex login --device-auth)");
        signInDeviceButton.addActionListener(e -> openSignInTerminal(true));

        JBLabel deviceAuthNote = new JBLabel(
            "<html>Use <i>headless</i> sign-in on remote/SSH machines where a browser cannot open automatically.</html>");
        deviceAuthNote.setForeground(UIUtil.getContextHelpForeground());
        deviceAuthNote.setFont(JBUI.Fonts.smallFont());

        mainPanel = FormBuilder.createFormBuilder()
            .addLabeledComponent("Binary:", binaryStatusLabel)
            .addLabeledComponent("Auth:", authStatusLabel)
            .addComponent(installNote, 2)
            .addComponent(installLink, 2)
            .addSeparator(8)
            .addLabeledComponent("Codex binary path:", binaryPathField)
            .addTooltip("Leave empty to auto-detect on PATH.")
            .addSeparator(8)
            .addComponent(signInButton, 4)
            .addComponent(signInDeviceButton, 2)
            .addComponent(deviceAuthNote, 2)
            .addComponentFillVertically(new JPanel(), 0)
            .getPanel();
        mainPanel.setBorder(JBUI.Borders.empty(8));
        return mainPanel;
    }

    @Override
    public boolean isModified() {
        if (binaryPathField == null) return false;
        String stored = nullToEmpty(loadCustomBinaryPath());
        return !binaryPathField.getText().trim().equals(stored);
    }

    @Override
    public void apply() {
        if (binaryPathField == null) return;
        saveCustomBinaryPath(binaryPathField.getText().trim());
    }

    @Override
    public void reset() {
        if (binaryPathField == null) return;
        refreshStatusAsync();
        binaryPathField.setText(nullToEmpty(loadCustomBinaryPath()));
    }

    @Override
    public void disposeUIResources() {
        binaryStatusLabel = null;
        authStatusLabel = null;
        binaryPathField = null;
        mainPanel = null;
    }

    // ── Status refresh ────────────────────────────────────────────────────────

    private void refreshStatusAsync() {
        if (binaryStatusLabel == null) return;
        binaryStatusLabel.setText(CHECKING);
        binaryStatusLabel.setForeground(UIUtil.getLabelForeground());
        authStatusLabel.setText(CHECKING);
        authStatusLabel.setForeground(UIUtil.getLabelForeground());

        ApplicationManager.getApplication().executeOnPooledThread(() -> {
            String customPath = loadCustomBinaryPath();
            String binary = customPath != null && !customPath.isEmpty() ? customPath : "codex";
            String version = BinaryDetector.detectBinaryVersion(binary, new String[0]);
            CodexCredentials creds = CodexCredentials.read();
            SwingUtilities.invokeLater(() -> applyStatusResults(version, creds));
        });
    }

    private void applyStatusResults(String version, CodexCredentials creds) {
        if (binaryStatusLabel == null) return;
        if (version != null) {
            binaryStatusLabel.setText("✓ Codex CLI found — " + version);
            binaryStatusLabel.setForeground(new Color(0, 128, 0));
        } else {
            binaryStatusLabel.setText("Codex CLI not found on PATH — install with npm install -g @openai/codex");
            binaryStatusLabel.setForeground(Color.RED);
        }

        if (authStatusLabel == null) return;
        if (creds.isLoggedIn()) {
            String name = creds.getDisplayName();
            authStatusLabel.setText("✓ Authenticated" + (name != null ? " — " + name : ""));
            authStatusLabel.setForeground(new Color(0, 128, 0));
        } else {
            authStatusLabel.setText("Not authenticated — run 'codex login' or click Sign In below");
            authStatusLabel.setForeground(Color.RED);
        }
    }

    // ── Sign-in helpers ───────────────────────────────────────────────────────

    /**
     * Opens a terminal tab and runs {@code codex login} (or {@code codex login --device-auth}
     * for headless/remote environments).
     */
    private void openSignInTerminal(boolean deviceAuth) {
        String customPath = loadCustomBinaryPath();
        String binary = customPath != null && !customPath.isEmpty() ? customPath : "codex";
        String cmd = deviceAuth ? binary + " login --device-auth" : binary + " login";
        AuthTerminalHelperKt.runAuthInEmbeddedTerminal(project, cmd, "Codex Sign In",
            () -> {
                openExternalTerminal(cmd);
                return kotlin.Unit.INSTANCE;
            });
    }

    private void openExternalTerminal(String cmd) {
        ApplicationManager.getApplication().executeOnPooledThread(() -> {
            try {
                String os = System.getProperty("os.name", "").toLowerCase();
                if (os.contains("win")) {
                    new ProcessBuilder("cmd", "/c", "start", "cmd", "/k", cmd).start();
                } else if (os.contains("mac")) {
                    new ProcessBuilder("osascript", "-e",
                        "tell application \"Terminal\" to do script \"" + cmd + "\"").start();
                } else {
                    new ProcessBuilder("sh", "-c",
                        "x-terminal-emulator -e '" + cmd + "' || gnome-terminal -- bash -c '" + cmd + "' || xterm -e bash -c '" + cmd + "'").start();
                }
            } catch (Exception ex) {
                SwingUtilities.invokeLater(() ->
                    JOptionPane.showMessageDialog(null,
                        "Could not open a terminal. Run manually: " + cmd,
                        "Codex Sign In", JOptionPane.INFORMATION_MESSAGE));
            }
        });
    }

    // ── Storage helpers (delegate to AcpClient persistence) ──────────────────

    private String loadCustomBinaryPath() {
        return com.github.catatafishen.ideagentforcopilot.acp.client.AcpClient
            .loadCustomBinaryPath(CodexAppServerClient.PROFILE_ID);
    }

    private void saveCustomBinaryPath(String path) {
        com.github.catatafishen.ideagentforcopilot.acp.client.AcpClient
            .saveCustomBinaryPath(CodexAppServerClient.PROFILE_ID, path);
    }

    @NotNull
    private static String nullToEmpty(String s) {
        return s != null ? s : "";
    }
}

File: plugin-core/src/main/java/com/github/catatafishen/agentbridge/settings/CodexClientConfigurable.kt
package com.github.catatafishen.agentbridge.settings

import com.github.catatafishen.agentbridge.acp.client.AcpClient
import com.github.catatafishen.agentbridge.services.AgentProfileManager
import com.github.catatafishen.agentbridge.ui.ThemeColor
import com.github.catatafishen.agentbridge.ui.runAuthInEmbeddedTerminal
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.options.BoundConfigurable
import com.intellij.openapi.options.SearchableConfigurable
import com.intellij.openapi.project.Project
import com.intellij.ui.HyperlinkLabel
import com.intellij.ui.JBColor
import com.intellij.ui.components.JBLabel
import com.intellij.ui.dsl.builder.AlignX
import com.intellij.ui.dsl.builder.bindItem
import com.intellij.ui.dsl.builder.bindText
import com.intellij.ui.dsl.builder.panel
import com.intellij.util.ui.UIUtil
import javax.swing.JOptionPane

@Suppress("unused")
class CodexClientConfigurable(private val project: Project) :
    BoundConfigurable("Codex"),
    SearchableConfigurable {

    private val statusLabel = JBLabel(CHECKING)

    override fun getId(): String = ID

    override fun createPanel() = panel {
        row("Binary:") { cell(statusLabel) }
        row {
            val installNote = JBLabel(
                "<html>Install with <code>npm install -g @openai/codex</code>, then run " +
                    "<code>codex login</code>. Authentication problems are reported by Codex " +
                    "itself when you send a prompt.</html>"
            )
            installNote.foreground = UIUtil.getContextHelpForeground()
            cell(installNote)
        }
        row {
            val link = HyperlinkLabel("Install Codex CLI — npmjs.com/@openai/codex")
            link.setHyperlinkTarget("https://www.npmjs.com/package/@openai/codex")
            cell(link)
        }
        separator()
        row("Codex binary path:") {
            textField()
                .align(AlignX.FILL)
                .resizableColumn()
                .applyToComponent { emptyText.text = "Auto-detect (leave empty)" }
                .comment("Leave empty to auto-detect on PATH.")
                .bindText(
                    { AgentProfileManager.getInstance().loadBinaryPath(PROFILE_ID).orEmpty() },
                    { AgentProfileManager.getInstance().saveBinaryPath(PROFILE_ID, it.trim()) }
                )
        }
        row("Bubble color:") {
            cell(ThemeColorComboBox())
                .comment("Choose a theme-aware accent color for message bubbles when using Codex.")
                .bindItem(
                    { ThemeColor.fromKey(AcpClient.loadAgentBubbleColorKey(PROFILE_ID)) },
                    { AcpClient.saveAgentBubbleColorKey(PROFILE_ID, it?.name) }
                )
        }
        separator()
        // Button labels include literal CLI commands ("codex login") which must stay lowercase.
        @Suppress("DialogTitleCapitalization")
        row { button("Sign in (codex login)") { openSignInTerminal(false) } }
        @Suppress("DialogTitleCapitalization")
        row { button("Sign in — headless (codex login --device-auth)") { openSignInTerminal(true) } }
        row {
            val note = JBLabel(
                "<html>Use <i>headless</i> sign-in on remote/SSH machines where a browser cannot " +
                    "open automatically.</html>"
            )
            note.foreground = UIUtil.getContextHelpForeground()
            cell(note)
        }
    }

    override fun reset() {
        super<BoundConfigurable>.reset()
        refreshStatusAsync()
    }

    private fun refreshStatusAsync() {
        statusLabel.text = CHECKING
        statusLabel.foreground = UIUtil.getLabelForeground()
        ApplicationManager.getApplication().executeOnPooledThread {
            val version = AcpClientBinaryResolver(PROFILE_ID, "codex").detectVersion()
            ApplicationManager.getApplication().invokeLater {
                if (version != null) {
                    statusLabel.text = "✓ Codex CLI found — $version"
                    statusLabel.foreground = JBColor(0x008000, 0x4EC94E)
                } else {
                    statusLabel.text =
                        "Codex CLI not found on PATH — install with npm install -g @openai/codex"
                    statusLabel.foreground = JBColor.RED
                }
            }
        }
    }

    private fun openSignInTerminal(deviceAuth: Boolean) {
        val customPath = AgentProfileManager.getInstance().loadBinaryPath(PROFILE_ID)
        val binary = if (!customPath.isNullOrEmpty()) customPath else "codex"
        val cmd = if (deviceAuth) "$binary login --device-auth" else "$binary login"
        runAuthInEmbeddedTerminal(project, cmd, "Codex Sign In") {
            openExternalTerminal(cmd)
        }
    }

    private fun openExternalTerminal(cmd: String) {
        ApplicationManager.getApplication().executeOnPooledThread {
            try {
                val os = System.getProperty("os.name", "").lowercase()
                when {
                    os.contains("win") ->
                        ProcessBuilder("cmd", "/c", "start", "cmd", "/k", cmd).start()
                    os.contains("mac") ->
                        ProcessBuilder(
                            "osascript", "-e",
                            "tell application \"Terminal\" to do script \"$cmd\""
                        ).start()
                    else -> ProcessBuilder(
                        "sh", "-c",
                        "x-terminal-emulator -e '$cmd' || gnome-terminal -- bash -c '$cmd' " +
                            "|| xterm -e bash -c '$cmd'"
                    ).start()
                }
            } catch (ex: Exception) {
                ApplicationManager.getApplication().invokeLater {
                    JOptionPane.showMessageDialog(
                        null,
                        "Could not open a terminal. Run manually: $cmd",
                        "Codex Sign In",
                        JOptionPane.INFORMATION_MESSAGE
                    )
                }
            }
        }
    }

    companion object {
        const val ID = "com.github.catatafishen.agentbridge.client.codex"
    }
}

private const val CHECKING = "Checking…"
// Mirrors CodexAppServerClient.PROFILE_ID (kept in sync; cannot reference the Java constant from a const val).
private const val PROFILE_ID = "codex"
